community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java

```diff
@@ -120,147 +140,90 @@ public PropertyItem get()
+            throw new IllegalStateException();
+        }
+    }
 
-            case INT:
+    private void ensureValuesLoaded()
+    {
+        if ( remainingBlocksToRead > 0 )
+        {
+            if ( valuesAfterHeader == null || valuesAfterHeader.length < remainingBlocksToRead )
             {
-                buffer.putInt( (int) header );
-                break;
+                valuesAfterHeader = new long[remainingBlocksToRead];
             }
 
-            case LONG:
+            int offset = cursor.getOffset();
+            try
             {
-                if ( (header & 0x1L) > 0 )
-                {
-                    buffer.putLong( header >>> 1 );
-                }
-                else
+                do
                 {
-                    buffer.putLong( cursor.getLong() );
-                    remainingBlocksToRead--;
+                    cursor.setOffset( offset );
+                    for ( int i = 0; remainingBlocksToRead-- > 0; i++ )
+                    {
+                        valuesAfterHeader[i] = cursor.getLong();
+                    }
                 }
-                break;
+                while ( cursor.shouldRetry() );
             }
-
-            case FLOAT:
+            catch ( IOException e )
             {
-                buffer.putInt( (int) header );
-                break;
+                throw new UnderlyingStorageException( e );
             }
+        }
+    }
 
-            case DOUBLE:
-            {
-                buffer.putLong( cursor.getLong() );
-                remainingBlocksToRead--;
-                break;
-            }
-
-            case STRING:
-            {
-                int storeOffset = cursor.getOffset();
-                cursor.close();
-
-                if ( stringRecordCursor == null )
-                {
-                    stringRecordCursor = stringStore.newDynamicRecordCursor();
-                }
-
-                try ( GenericCursor<DynamicRecord> stringRecords = stringStore.getRecordsCursor( header, true,
-                        stringRecordCursor ) )
-                {
-                    while ( stringRecords.next() )
-                    {
-                        DynamicRecord dynamicRecord = stringRecords.get();
-
-                        buffer.put( dynamicRecord.getData(), 0, dynamicRecord.getData().length );
-                    }
-                }
+    private void readFromStore( AbstractDynamicStore store ) throws IOException
+    {
+        int storeOffset = cursor.getOffset();
+        cursor.close();
+        cursor = null;
 
-                cursor = propertyStore.newReadCursor( currentRecordId );
-                cursor.setOffset( storeOffset );
-                break;
-            }
+        if ( recordCursor == null )
+        {
+            recordCursor = store.newDynamicRecordCursor();
+        }
 
-            case ARRAY:
+        buffer.clear();
+        long startBlockId = (header & 0xFFFFFFFFF0000000L) >>> 28;
```

Comment:
We already have `org.neo4j.kernel.impl.store.record.PropertyBlock` that already know how to do shift and mask operations. That should be used here